pilot/pkg/serviceregistry/serviceentry/servicediscovery.go

```diff
@@ -503,7 +505,17 @@ func (s *ServiceEntryStore) Services() ([]*model.Service, error) {
 	s.mutex.RLock()
 	allServices := s.services.getAllServices()
 	s.mutex.RUnlock()
-	return autoAllocateIPs(allServices), nil
+
+	autoAllocateIPs(allServices)
+	out := make([]*model.Service, len(allServices))
+	for _, svc := range allServices {
+		// TODO: eliminate the shallowcopy here
+		// autoAllocateIPs will re-allocate ips for the service,
+		// if return the pointer directly, there will be a race with `BuildNameTable`
+		tmp := *svc
+		out = append(out, &tmp)
```

Comment:
https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35542/unit-tests_istio/1458711022863912960

This is for the race test. I think we need to improve later to prevent allocate again for all the services